bench/workgen/workgen.cxx

```diff
@@ -362,29 +320,59 @@ WorkloadRunner::start_dynamic_table_mgmt(WT_CONNECTION *conn)
                 /* Walk the list and select the first that is not already in delete list */
                 for (auto const &it : icontext->_dyn_tint) {
-                    if (std::find(icontext->_dyn_tables_delete.begin(),
-                          icontext->_dyn_tables_delete.end(),
-                          it.first) != icontext->_dyn_tables_delete.end()) {
+                    if (icontext->_dyn_table_runtime[it.second]._pending_delete) {
                         continue;
                     } else {
                         std::cout << "Marking pending removal for " << it.first << "\n";
-                        icontext->_dyn_tables_delete.push_back(it.first);
+                        icontext->_dyn_table_runtime[it.second]._pending_delete = true;
+                        tables_delete.push_back(it.first);
                         break;
                     }
                 }
             }
+
+            /* Process any pending delete, the actual table drop will be done without holding the
+             * lock */
+            for (size_t i = 0; i < tables_delete.size();) {
+                // The table might still be in use.
+                const std::string uri(tables_delete.at(i));
+                tint_t tint = icontext->_dyn_tint.at(uri);
+                if (icontext->_dyn_table_runtime[tint]._in_use != 0) {
+                    std::cout << "Inuse " << uri << "\n";
+                    ++i;
+                    continue;
+                }
+
+                // Delete all local data related to the table.
+                tables_delete.erase(tables_delete.begin() + i);
+                icontext->_dyn_tint.erase(uri);
+                icontext->_dyn_table_names.erase(tint);
+                icontext->_dyn_table_runtime.erase(tint);
+                drop_uris.push_back(uri);
+            }
+        }
+
+        /*
+         * We will drop the WiredTiger tables without holding the lock. These tables have been
+         * removed from the shared data structures, and we know no thread is operating on them.
+         */
+        for (auto uri : drop_uris) {
+            /* Spin on EBUSY, we do not expect to get stuck */
+            while (
+              (ret = session->drop(session, uri.c_str(), "force,checkpoint_wait=false")) == EBUSY) {
+                std::cout << "Drop returned EBUSY for table " << uri << "\n";
+            }
+            if (ret != 0)
+                THROW("Table drop failed in start_dynamic_table_mgmt.");
+
+            std::cout << "Dropped table " << uri << "\n";
         }
+        drop_uris.clear();
 
-        /* Garbage collect - delete tables pending delete and no longer in use */
-        if (create_count % 6 == 0)
-            garbage_collection(icontext, session);
         sleep(1);
     }
 
-    // Sleep a bit before existing and garbage collect
-    sleep(5);
-    garbage_collection(icontext, session);
-
+    // Do we need to cleanup tables that might be pending delete when stopping
```

Comment:
I don't think this is necessary but we could add a warning message in the output.